ql/src/java/org/apache/hadoop/hive/ql/lockmgr/DbTxnManager.java

```diff
@@ -1163,9 +1163,15 @@ private static class MaterializationRebuildLockHeartbeater implements Runnable {
     public void run() {
       LOG.trace("Heartbeating materialization rebuild lock for {} for query: {}",
           AcidUtils.getFullTableName(dbName, tableName), queryId);
-      boolean refreshed;
+      boolean refreshed = false;
       try {
-        refreshed = txnMgr.heartbeatMaterializationRebuildLock(dbName, tableName, txnId);
+        if (!txnMgr.getValidTxns().isTxnAborted(txnId)) {
```

Comment:
Selecting all the transactions in a heartbeat, just to check if the txn is aborted looks like an unnecessary resource consumption that we could avoid.

Reply:
Changed now as per @kasakrisz suggestion.